File: library/kotlin/io/envoyproxy/envoymobile/RequestBuilder.kt
package io.envoyproxy.envoymobile

import java.net.URL


/**
 * Builder used for constructing instances of `Request` types.
 *
 * @param url URL for the request.
 * @param method Method for the request.
 */
class RequestBuilder(
    val url: URL,
    val method: RequestMethod
) {
  // Headers to send with the request.
  // Multiple values for a given name are valid, and will be sent as comma-separated values.
  private val headers: MutableMap<String, MutableList<String>> = mutableMapOf()

  // Trailers to send with the request.
  // Multiple values for a given name are valid, and will be sent as comma-separated values.
  private val trailers: MutableMap<String, MutableList<String>> = mutableMapOf()

  // Serialized data to send as the body of the request.
  private var body: ByteArray? = null

  // Retry policy to use for this request.
  private var retryPolicy: RetryPolicy? = null

  fun addBody(body: ByteArray?): RequestBuilder {
    this.body = body
    return this
  }

  fun addRetryPolicy(retryPolicy: RetryPolicy?): RequestBuilder {
    this.retryPolicy = retryPolicy
    return this
  }

  fun addHeader(header: String, value: String): RequestBuilder {
    if (headers.containsKey(header)) {
      headers[header]!!.add(value)
    }

    headers[header] = mutableListOf(value)
    return this
  }

  fun removeHeader(header: String): RequestBuilder {
    headers.remove(header)
    return this
  }

  fun addTrailer(trailer: String, value: String): RequestBuilder {
    if (trailers.containsKey(value)) {
      trailers[trailer]!!.add(value)
    }

    trailers[trailer] = mutableListOf(value)
    return this
  }

  fun removeTrailer(trailer: String): RequestBuilder {
    trailers.remove(trailer)
    return this
  }

  fun build(): Request {
    return Request(
        method,
        url,
        headers,
        trailers,
        body,
        retryPolicy
    )
  }

  internal fun setHeaders(headers: Map<String, List<String>>): RequestBuilder {
    this.headers.clear()
    headers.map { entry -> this.headers[entry.key] = entry.value.toMutableList() }
    return this
  }

  internal fun setTrailers(trailers: Map<String, List<String>>): RequestBuilder {
    this.trailers.clear()
    trailers.map { entry -> this.trailers[entry.key] = entry.value.toMutableList() }
    return this
  }
}

File: library/swift/src/RequestBuilder.swift
import Foundation

/// Builder used for constructing instances of `Request` types.
@objcMembers
public final class RequestBuilder: NSObject {
  /// Method for the request.
  public let method: RequestMethod
  /// The URL scheme for the request (i.e., "https").
  public let scheme: String
  /// The URL authority for the request (i.e., "api.foo.com").
  public let authority: String
  /// The URL path for the request (i.e., "/foo").
  public let path: String
  /// The protcol version to use for upstream requests.
  public let upstreamHttpProtocol: UpstreamHttpProtocol
  /// Headers to send with the request.
  /// Multiple values for a given name are valid, and will be sent as comma-separated values.
  public private(set) var headers: [String: [String]] = [:]
  // Retry policy to use for this request.
  public private(set) var retryPolicy: RetryPolicy?

  // MARK: - Initializers

  /// Internal initializer used for converting a request back to a builder.
  init(request: Request) {
    self.method = request.method
    self.scheme = request.scheme
    self.authority = request.authority
    self.path = request.path
    self.upstreamHttpProtocol = request.upstreamHttpProtocol
    self.headers = request.headers
    self.retryPolicy = request.retryPolicy
  }

  /// Public initializer.
  public init(method: RequestMethod,
              scheme: String = "https",
              authority: String,
              path: String,
              upstreamHttpProtocol: UpstreamHttpProtocol)
  {
    self.method = method
    self.scheme = scheme
    self.authority = authority
    self.path = path
    self.upstreamHttpProtocol = upstreamHttpProtocol
  }

  // MARK: - Builder functions

  @discardableResult
  public func addHeader(name: String, value: String) -> RequestBuilder {
    self.headers[name, default: []].append(value)
    return self
  }

  @discardableResult
  public func removeHeaders(name: String) -> RequestBuilder {
    self.headers.removeValue(forKey: name)
    return self
  }

  @discardableResult
  public func removeHeader(name: String, value: String) -> RequestBuilder {
    self.headers[name]?.removeAll(where: { $0 == value })
    if self.headers[name]?.isEmpty == true {
      self.headers.removeValue(forKey: name)
    }

    return self
  }

  @discardableResult
  public func addRetryPolicy(_ retryPolicy: RetryPolicy) -> RequestBuilder {
    self.retryPolicy = retryPolicy
    return self
  }

  public func build() -> Request {
    return Request(method: self.method,
                   scheme: self.scheme,
                   authority: self.authority,
                   path: self.path,
                   upstreamHttpProtocol: self.upstreamHttpProtocol,
                   headers: self.headers,
                   retryPolicy: self.retryPolicy)
  }
}

// MARK: - Objective-C helpers

extension Request {
  /// Convenience builder function to allow for cleaner Objective-C syntax.
  ///
  /// For example:
  ///
  /// Request *req = [Request withMethod:RequestMethodGet (...) build:^(RequestBuilder *builder) {
  ///   [builder addHeaderWithName:@"x-some-header" value:@"foo"];
  /// }];
  @objc
  public static func with(method: RequestMethod,
                          scheme: String,
                          authority: String,
                          path: String,
                          upstreamHttpProtocol: UpstreamHttpProtocol,
                          build: (RequestBuilder) -> Void)
    -> Request
  {
    let builder = RequestBuilder(method: method, scheme: scheme,
                                 authority: authority, path: path, upstreamHttpProtocol: upstreamHttpProtocol)
    build(builder)
    return builder.build()
  }
}
